pokemongo_bot/walkers/polyline_walker.py

```diff
@@ -1,29 +1,37 @@
-from random import uniform
+from geographiclib.geodesic import Geodesic
 
-from pokemongo_bot.cell_workers.utils import distance
 from pokemongo_bot.walkers.step_walker import StepWalker
 from polyline_generator import PolylineObjectHandler
 
 
 class PolylineWalker(StepWalker):
+    def get_next_position(self, origin_lat, origin_lng, origin_alt, dest_lat, dest_lng, dest_alt, distance):
+        polyline = PolylineObjectHandler.cached_polyline((self.bot.position[0], self.bot.position[1]), (dest_lat, dest_lng), distance, google_map_api_key=self.bot.config.gmapkey)
 
-    def __init__(self, bot, dest_lat, dest_lng):
-        self.bot = bot
-        self.speed = uniform(self.bot.config.walk_min, self.bot.config.walk_max)
-        self.dest_lat, self.dest_lng = dest_lat, dest_lng
-        self.actual_pos = tuple(self.bot.position[:2])
-        self.actual_alt = self.bot.position[-1]
-        self.polyline = PolylineObjectHandler.cached_polyline(self.actual_pos,
-                                                              (self.dest_lat, self.dest_lng),
-                                                              self.speed, google_map_api_key=self.bot.config.gmapkey)
-        self.pol_lat, self.pol_lon = self.polyline.get_pos()
-        self.pol_alt = self.polyline.get_alt() or self.actual_alt
-        super(PolylineWalker, self).__init__(self.bot, self.pol_lat, self.pol_lon,
-                                             self.pol_alt, fixed_speed=self.speed)
-
-    def step(self):
-        step = super(PolylineWalker, self).step()
-        if not (distance(self.pol_lat, self.pol_lon, self.dest_lat, self.dest_lng) > 10 and step):
-            return False
-        else:
-            return True
+        while True:
+            _, (dest_lat, dest_lng) = polyline._step_dict[polyline._step_keys[polyline._last_step]]
+            dest_alt = polyline.get_alt() or origin_alt
+
+            next_lat, next_lng, next_alt = super(PolylineWalker, self).get_next_position(origin_lat, origin_lng, origin_alt, dest_lat, dest_lng, dest_alt, distance)
+
+            if polyline._last_step == len(polyline._step_keys) - 1:
+                break
+            else:
+                travelled = Geodesic.WGS84.Inverse(origin_lat, origin_lng, next_lat, next_lng)["s12"]
+                remaining = Geodesic.WGS84.Inverse(next_lat, next_lng, dest_lat, dest_lng)["s12"]
+                step_distance = Geodesic.WGS84.Inverse(origin_lat, origin_lng, dest_lat, dest_lng)["s12"]
```

Comment:
BTW, step_distance is the dictionary key the one you are discarding at line 12:

`_, (dest_lat, dest_lng) = polyline._step_dict[polyline._step_keys[polyline._last_step]]`